### PR TITLE
Userlist

### PR DESCRIPTION
**Pull request by borisyankov**
* Flatten userlist store structure
* Extract userlist sort and filter as selectors
* Merge presense response with userlist state
* Better detection of who is active/offline, exclude self from list
* Unit tests
* Wallaby.js config

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@borisyankov, it looks like you've already signed the Dropbox CLA. Thanks!

File: src/userlist/__tests__/userListReducers-test.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        pushable: false,
+      },
+    },
+  };
+  const prevState = fromJS([{
+    full_name: 'Some Guy',
+    email: 'some@email.com',
+    status: 'offline',
+  }, {
+    full_name: 'Boris Yankov',
```
Comment (neerajwahi) [inline src/userlist/__tests__/userListReducers-test.js]: I'd prefer to use fake names and emails

File: src/userlist/userListReducers.js [hunk 2]
```diff
@@ -31,46 +26,57 @@ type PresenceMap = {
...
+        if (userIndex === -1) return currentState;
+
+        const p = action.presence[email];
+        const timestamp = timestampFromPresence(p);
+        const status = activityFromTimestamp(activityFromPresence(p), timestamp);
+        return currentState.setIn([userIndex, 'status'], status).setIn([userIndex, 'timestamp'], timestamp);
+      }, state);
+    }
+    case GET_USER_RESPONSE: {
+      const users = action.users.map(x => ({
```
Comment (neerajwahi) [inline src/userlist/userListReducers.js]: Can we rename `x` to `user`? I think it's clearer to avoid short variable names like `x` unless it's a loop index or a really short function

File: src/userlist/userListSelectors.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  })[status]);
+
+export const sortUserList = (users: any[]): any[] =>
+  users.sort((x1, x2) =>
+    statusOrder(x1.get('status')) - statusOrder(x2.get('status')) ||
+    x1.get('fullName').toLowerCase().localeCompare(x2.get('fullName').toLowerCase())
+  );
+
+export const filterUserList = (users: any[], filter: string = '', ownEmail: string): any[] =>
+  users.filter(x =>
```
Comment (neerajwahi) [inline src/userlist/userListSelectors.js]: Ditto above

File: src/userlist/__tests__/userListSelectors-test.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import { sortUserList, filterUserList } from '../userListSelectors';
+
+it('empty input to filterUserList results in empty list', () => {
+  const users = fromJS([]);
+  const filteredUsers = filterUserList(users, 'some filter');
+  expect(filteredUsers).toEqual(fromJS([]));
+});
+
+it('filterUserList returns same list if no filter', () => {
+  const users = fromJS([{ email: 'user1@email.com' }, { email: 'user2@email.com' }]);
```
Comment (neerajwahi) [inline src/userlist/__tests__/userListSelectors-test.js]: Let's use example.com throughout the codebase if you don't mind. It's been established for this purpose :)

File: src/userlist/userListReducers.js [hunk 5]
```diff
@@ -31,46 +26,57 @@ type PresenceMap = {
...
-        const status = activityFromPresence(action.presence[x]);
-        const user = state.find(u => u.email === x);
-        newState = state.setIn([user, { status }]);
-      });
-      return newState;
+      return Object.keys(action.presence).reduce((currentState, email) => {
+        const userIndex = state.findIndex(u => u.get('email') === email);
+        if (userIndex === -1) return currentState;
+
+        const p = action.presence[email];
```
Comment (neerajwahi) [inline src/userlist/userListReducers.js]: Can we rename this to something like `presence`? `p` isn't descriptive

**Comment by neerajwahi** [comment]
Thanks @borisyankov.

One other comment: mind making the git commit message titles shorter? You could put any extra details inside of the message (after the first line). And if a commit covers multiple unrelated changes, we should split it up into multiple commits if possible.

File: src/userlist/__tests__/userListReducers-test.js (hunk 1 above)
Comment (borisyankov) [inline src/userlist/__tests__/userListReducers-test.js]: Sure. I was experimenting with parts of real responses. :)

File: src/userlist/__tests__/userListReducers-test.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
+import { fromJS } from 'immutable';
+import { GET_USER_RESPONSE, PRESENCE_RESPONSE } from '../userListActions';
+import userListReducers, { activityFromPresence, timestampFromPresence } from '../userListReducers';
+
+it('when unrecognized action, no previous state, returns initial state, does not throw', () => {
```
Comment (neerajwahi) [inline src/userlist/__tests__/userListReducers-test.js]: These test descriptions are a little hard to parse -- maybe we could rephrase them? Such as `it('returns initial state and does not throw for unknown action with no previous state')`..

I find it a little easier to read when the result comes first, followed by the conditions, but that may just be a personal preference

File: src/userlist/UsersCard.js [hunk 7]
```diff
@@ -32,6 +34,10 @@ export default class UsersCard extends Component {
     filter: string,
   };
 
+  static contextTypes = {
+    drawer: () => null,
```
Comment (neerajwahi) [inline src/userlist/UsersCard.js]: We probably want to make the drawer's open/close status part of the app's state eventually

File: src/userlist/userListActions.js [hunk 8]
```diff
@@ -21,3 +20,9 @@ export const sendGetUsers = (auth: Auth) =>
       users: response,
     });
   };
+
+export const sendInitialGetUsers = (auth: Auth) =>
```
Comment (neerajwahi) [inline src/userlist/userListActions.js]: Why name actions with the `send` prefix? I find it superfluous personally (and maybe a little confusing)

File: src/userlist/userListSelectors.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+const statusOrder = (status) =>
```
Comment (neerajwahi) [inline src/userlist/userListSelectors.js]: This could be more readable. Maybe something like the following?
```
const statusOrder = (status) =>
    case (status) {
        active:
            return 1;
        idle:
            return 2;
...
```

File: src/userlist/UsersCard.js (hunk 7 above)
Comment (borisyankov) [inline src/userlist/UsersCard.js]: It is either going to be part of the Navigation state, which is WIP or just a temporary component state.

Any state that does not need to be accessed from anywhere else, is recommended (officially) to not be in Redux, but be in the component state.

File: src/userlist/userListActions.js (hunk 8 above)
Comment (borisyankov) [inline src/userlist/userListActions.js]: This is to differentiate from the action and the api call called getUsers. We can use other names, but they should definitely differ.

**Comment by borisyankov** [comment]
Updated.
